Add support for layering local RPMs

jlebon opened the pull request.

jlebon commented:
This still needs some playing around and of course tests!

jlebon commented:
There's also a couple of design decisions that we might want to revisit. E.g. right now, we import into the same cache as packages obtained from other sources (but we do use SHA-256 to verify that it hasn't changed), though we might want to put those under a different namespace (e.g. `rpmostree/pkglocal/`) or even repo.

cgwalters commented on src/app/rpmostree-pkg-builtins.c:
```diff
@@ -157,7 +158,37 @@ rpmostree_builtin_pkg_add (int            argc,
+        }
+
+      /* OK, it's a relative path, we need to check that it
+       * exists and canonicalize it for the daemon. */
```
Can we open it here and pass the fd over dbus?

jlebon replied:
Nice, I didn't realize D-Bus supported this. I'd like to try it out, though it's a bit unfortunate that it'd have to be hidden away in the `options` array if we want to avoid adding another method.

jlebon replied:
Mind if we do this in a separate PR?

cgwalters replied:
That's fine yep.

jlebon replied:
https://github.com/projectatomic/rpm-ostree/pull/696

cgwalters commented on src/libpriv/rpmostree-core.c:
```diff
@@ -670,6 +690,227 @@ on_hifstate_percentage_changed (DnfState   *hifstate,
+      const char *ref = k;
+      g_autofree char *cache_nevra = cache_branch_to_nevra (ref);
+
+      if (g_str_equal (nevra, cache_nevra))
```
minor: Could avoid rightward shift here by doing `if (!equal()) continue;`

cgwalters commented on src/libpriv/rpmostree-core.c:
```diff
@@ -727,6 +968,47 @@ append_quoted (GString *r, const char *value)
 
+/* translates cachebranch back to nevra */
+static char *
+cache_branch_to_nevra (const char *cachebranch)
```
This is the type of thing that wants inline unit tests.  Not necessarily something to do in this patch, just noting in passing.

jlebon replied:
Added tests below!

cgwalters commented on src/libpriv/rpmostree-core.c:
```diff
@@ -727,6 +968,47 @@ append_quoted (GString *r, const char *value)
+
+  for (p = cachebranch; *p; p++)
+    {
+      if (*p != '_')
```
I'd `char c = *p` and use that down the line here.

cgwalters commented on src/libpriv/rpmostree-core.c:
```diff
@@ -1011,31 +1311,51 @@ rpmostree_context_prepare_install (RpmOstreeContext    *self,
+      if (pkg_to_checksum && g_hash_table_contains (pkg_to_checksum, pkg))
+        {
+          /* the ostree checksum itself is already a good representation
+           * of the whole package */
```
Hmm...but it includes a timestamp.  We should probably make the commit timestamp derive from some timestamp in the rpm header?

jlebon replied:
Good point. Done below!

cgwalters commented on src/libpriv/rpmostree-core.c:
```diff
@@ -1842,8 +2171,10 @@ run_pre_sync (int tmp_metadata_dfd,
 {
   _cleanup_rpmheader_ Header hdr = NULL;
+  g_autofree char *path =
+    g_strdup_printf ("%s.rpm", dnf_package_get_nevra (pkg));
```
At this point, we should make this a function I'd say.

cgwalters commented on src/libpriv/rpmostree-core.c:
```diff
@@ -1998,6 +2331,23 @@ checksum_version (GVariant *checksum)
+  DnfPackage *p1 = (DnfPackage*)v1;
+  DnfPackage *p2 = (DnfPackage*)v2;
+  return g_str_equal (dnf_package_get_package_id (p1),
+                      dnf_package_get_package_id (p2));
```
`dnf_pkg_cmp() == 0`?

jlebon replied:
Hmm, though that only compares the NEVRA, whereas `get_package_id` includes repo information. E.g. a package provided locally should still be considered distinct from a repo package, even if they have the same NEVRA. I think it wouldn't make a difference in this context, since by the time we add to that array, we've already made our determination, though it's a nice property to keep I think (and it's cached too, thankfully).

cgwalters replied:
Hum...should we be comparing e.g. `dnf_package_get_chksum()`?

jlebon replied:
So, the reason why all this was needed in the first place was because `dnf_package_get_chksum()` didn't work on our dummy RPMs (which is what's used in `rpmostree_dnf_add_checksum_goal()`). I decided to dig deeper as to why that was, which lead to https://github.com/rpm-software-management/libdnf/pull/273. Once this is merged, we can back out a lot of the logic there.

cgwalters commented on src/libpriv/rpmostree-unpacker.c:
```diff
@@ -473,6 +477,17 @@ build_metadata_variant (RpmOstreeUnpacker *self,
                                                      metadata, TRUE));
+
+    /* NB: the header also includes checksums for every entry in the (cut-off)
+     * archive, so there's no need to also bring that in */
```
See also https://github.com/rpm-software-management/rpm/issues/163

We should probably start defining terms more carefully since "pkg_checksum" could be like 3-4 different things 😄 
How about `header_sha256` or something?

cgwalters commented on src/libpriv/rpmostree-unpacker.c:
```diff
@@ -473,6 +477,17 @@ build_metadata_variant (RpmOstreeUnpacker *self,
+    g_checksum_update (pkg_checksum, g_bytes_get_data (metadata, NULL),
+                                     g_bytes_get_size (metadata));
+
+    self->hdr_sha256 = g_strdup (g_checksum_get_string (pkg_checksum));
```
Oh actually you call it that here.

cgwalters commented on src/libpriv/rpmostree-core.c:
```diff
@@ -670,6 +690,227 @@ on_hifstate_percentage_changed (DnfState   *hifstate,
+                                            &commit, NULL, error))
+                return FALSE;
+
+              if (!get_commit_header_sha256 (commit, &actual_sha256, error))
```
Do we need to handle the case where this doesn't exist?

jlebon replied:
Hmm, I don't *think* so. We still want to error out, since we can't verify that it's the same package we want. We should be fine from the backwards compatibility perspective too, since we would only request a match if we've previously been asked to install a local RPM (and we always re-unpack them), so the header would get created.

jlebon commented:
Rebased with fixups and tests added! ⬆️

jlebon commented:
Good ol' `compose` flake.

cgwalters commented on src/app/rpmostree-pkg-builtins.c:
```diff
@@ -157,7 +158,37 @@ rpmostree_builtin_pkg_add (int            argc,
+      g_autofree char *abspath = NULL;
+
+      if (!g_str_has_suffix (pkgspec, ".rpm") || /* repo install */
+           g_str_has_prefix (pkgspec, "/"))      /* local install */
```
How about `./` as an allowable prefix too?

jlebon replied:
That's a relative path though, right? So it'll be made absolute down below.

cgwalters commented on src/daemon/rpmostreed-transaction-types.c:
```diff
@@ -666,9 +722,46 @@ deploy_transaction_execute (RpmostreedTransaction *transaction,
-        goto out;
+      GHashTable *pkgs = g_hash_table_new_full (g_str_hash, g_str_equal,
+                                                g_free, NULL);
+      GHashTable *local_pkgs = g_hash_table_new_full (g_str_hash, g_str_equal,
```
Don't these need `g_autoptr()`?

cgwalters commented:
Was just playing with the branch; I wanted to see the behavior when layering an RPM that's already in the base.  I got in this situation:

```
[root@localhost ~]# rpm-ostree install ~vagrant/sync/src/distgit/fedora/ostree/x86_64/ostree-2016.14-1.fc26.x86_64.rpm 
Checking out tree bdd3eb7... done
error: Failed to find cached pkg for ostree-2017.2-4.fc26.x86_64
[root@localhost ~]# 
[root@localhost ~]# rpm-ostree install ~vagrant/sync/src/distgit/fedora/ostree/x86_64/ostree-2017.2-4.fc26.x86_64.rpm 
error: Package 'ostree-2017.2-4.fc26.x86_64' is already layered
[root@localhost ~]# rpm-ostree status
State: idle
Deployments:
  fahc:fahc/25/x86_64/buildmaster
             Version: 25.21 (2017-03-08 02:27:31)
          BaseCommit: bdd3eb7b5d7715e45d65c7f5f841612757634cf06a8ce367f2fa3af78db5e9f5
              Commit: 097721317de22d9b38c83720d5c6244d9e2da41b55a6488a2f8f4e8de8ecd796
              OSName: fedora-atomic
     LayeredPackages: cowsay fuse-sshfs
       LocalPackages: ostree-2017.2-4.fc26.x86_64

● fahc:fahc/25/x86_64/buildmaster
             Version: 25.21 (2017-03-07 22:20:10)
          BaseCommit: bdd3eb7b5d7715e45d65c7f5f841612757634cf06a8ce367f2fa3af78db5e9f5
              Commit: 0b220457235cbe7d965ffacb4520de09fbebd23522bc88c0d1869aaa36673fe5
              OSName: fedora-atomic
     LayeredPackages: cowsay fuse-sshfs
            Unlocked: development
[root@localhost ~]# 
```

I noticed the initial installation doesn't add a pkgcache ref.

cgwalters commented:
So more directly:
```
[root@localhost ~]# rpm-ostree install ~vagrant/sync/src/distgit/fedora/ostree/x86_64/ostree-2017.2-4.fc26.x86_64.rpm 
...
[root@localhost ~]# ostree --repo=/ostree/repo/extensions/rpmostree/pkgcache/ refs |grep /ostr|wc -l
0
```

jlebon commented:
OK, updated with some more fixes:
- Bumped libdnf to include https://github.com/rpm-software-management/libdnf/pull/273 and backed out changes that were previous required to work around that.
- Made the upgrader detect cases where the user tries to locally layer a NEVRA that's already in the base.
- As a follow-up to the above, made the core smarter about solutions that would require mutating the base layer. I spent too much time trying to get libsolv to do this natively for us, but AFAICT there is no such thing. So for now we just analyze the solution ourselves and give an appropriate error. Will try to research and reach out though to see if this is possible.

This fixes other issues that were present from before. E.g. if you had `foo-1.0-1.x86_64` installed and did `rpmostree install 'foo > 1.0'` (assuming such a pkg existed in the repos), you'd get the error:

```
error: No packages in installation set
```

Which was because the pkg foo was marked for UPGRADE, and thus wasn't caught by the assemble's query (which only greps for INSTALL). Instead, now you get:

```
error: The following packages would replace a base package: foo-0.1-1.x86_64
```

jlebon commented:
> So more directly:
> ```
> [root@localhost ~]# rpm-ostree install ~vagrant/sync/src/distgit/fedora/ostree/x86_64/ostree-2017.2-4.fc26.x86_64.rpm

This should error out now that we actually check for mutation in the base. I think what happened there was that you already had packages layered, so didn't hit the `No packages in installation set` error up above.

cgwalters commented on src/daemon/rpmostree-sysroot-upgrader.c:
```diff
@@ -549,6 +549,25 @@ generate_treespec (GHashTable *packages,
 }
 
+static gboolean
+sack_has_subject (DnfSack    *sack,
```
Ah, nice.  This approach seems like it's going to solve a bunch of problems indeed.

cgwalters commented on src/daemon/rpmostree-sysroot-upgrader.c:
```diff
@@ -614,27 +633,26 @@ find_missing_pkgs_in_rpmdb (int            rootfs_dfd,
+          if (sack_has_subject (rsack->sack, nevra))
+            {
+              g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                           "Package '%s' is already in the base.", nevra);
```
No periods at the end of errors by convention.  Having one here is an error

cgwalters commented on src/libpriv/rpmostree-core.c:
```diff
@@ -1353,39 +1350,23 @@ rpmostree_context_prepare_install (RpmOstreeContext    *self,
+      int pkg_checksum_type;
+      const unsigned char *pkg_checksum_bytes = dnf_package_get_chksum (pkg, &pkg_checksum_type);
+      g_autofree char *pkg_checksum = hy_chksum_str (pkg_checksum_bytes, pkg_checksum_type);
+      char *pkg_checksum_with_type = g_strconcat (hy_chksum_name (pkg_checksum_type), ":", pkg_checksum, NULL);
```
Don't need to change this, but I tend to prefer using `g_autoptr()` always, and then use `g_steal_pointer()` if we're going to move.  It's more robust against future code refactoring if e.g. one adds an error exit before the ownership transfer.

jlebon replied:
Yeah, I definitely agree. Though I left it as is -- that fixup is actually just restoring code that my initial commits replaced. I opted to just let it be to minimize the diff and make git blame less noisy.

cgwalters commented:
OK, very nice work here (as usual).   Just a stream of minor nits.  I think it makes sense to land this in git master (and hence {F,C}AHC) have people stress test it for a bit.  We just did a release, so let's think of this as the start of a new cycle.

Hence I'm ready to r+ after those minor fixups.

jlebon commented:
OK, fixed msg and rebased onto master! (Actually, git was able to 3-way merge on top of #670, but I wanted to make doubly sure we did the same thing for the new `requested-local-packages` element).

cgwalters commented:
@rh-atomic-bot r+ 84e4154

🎉
